src/database/DataTypes/Transaction.js

```diff
@@ -121,6 +120,49 @@ export class Transaction extends Realm.Object {
   }
 
   /**
+   * Remove transaction batches with given ids from this transaction
+   * Also, remove ItemBatch associated  with TransactionBatch if this transaction
+   * is external supplier invoice, and remove all TransactionItems from transaction
+   * that don't have TransactionBatch
+   * @param  {Realm}  database  App wide local database
+   * @param  {array}  transactionBatchIds The ids of transactionBatches to remove
+   * @return {none}
+   */
+  removeTransactionBatchesById(database, transactionBatchIds) {
+    if (this.isFinalised) throw new Error('Cannot modify finalised transaction');
+    if (!this.isValid()) return;
+
+    const transactionBatches = this.getTransactionBatches(database);
+    transactionBatchIds.forEach(transactionBatchId => {
+      const transactionBatch = transactionBatches.find(matchTransactionBatch =>
+                                  matchTransactionBatch.id === transactionBatchId);
+      // Can only safe remove ItemBatch if it's been created by ExternalSupplierInvoice
+      if (this.isExternalSupplierInvoice) database.delete('ItemBatch', transactionBatch.itemBatch);
+      database.delete('TransactionBatch', transactionBatch);
+    });
+
+    this.pruneBatchlessTransactionItems(database);
+  }
+
+  /**
+   * Removes this transaction and associated TransactionBatches and TransactionItems
+   * Also, removes ItemBatches associated  with transaction if transaction
+   * is external supplier invoice
+   * @param  {Realm}  database  App wide local database
+   * @return {none}
+   */
+  removeSelf(database) {
+    if (this.isFinalised) throw new Error('Cannot delete finalised transaction');
+    if (!this.isValid()) return;
+    const transactionBatchesIds =
+      this.getTransactionBatches(database).map(transactionBatch => transactionBatch.id);
+
+    this.removeTransactionBatchesById(database, transactionBatchesIds);
+
+    database.delete('Transaction', this);
```

Comment:
This should move into the destructor, so you can always just call `database.delete` and it will clean up all of the batches etc.